### ISSUE TITLE
Indefinite retries of failing objects

### ISSUE THREAD

**Issue by brendandburns** [kubernetes/kubernetes#9382]
From: _Uday Kumar Raju ruddarraju <rudaykumarraju@gmail.com>_

I have run into a wierd situation using the FIFO queue in the ServiceController. We (at ebay) have modified the openstack cloud provider to suit our installation of openstack, where all the Pod IPs are reachable outside the cluster. So we have changed out cloudprovider to read the pods belonging to a service and create a tcp load balancer with pod IPs as the members. We saw a wierd case recently with the way ListWatcher is enabled.

Here is the pseudo code of our CreateTCPLoadbalacer:

```go
func CreateTCPLoadbalancer(lbName) {
  GetAllPodsUnderService() // A call to API server to get all the pods with the same selector as the service
  CallOpenstackToCreateLBPool(PodIPs, method)
}
```

You can see this is a little different from all the public providers where, none of the other providers have a dependency back to the apiserver during this operation.

I have run into a situation where servicecontroller indefinitely tries to call the CreateTCPLoadbalancer even when it fails everytime. My failure case happened to be this:

- Called Create Service for a set of Pods. Created a delta on the fifo queue
- The CreateTCPLoadbalancer call failed due to a connection reset on the Openstack endpoint due to a network issue, so the event has been put back on the FIFO queue as the CreateTCPLoadbalancer failure is always considered to be a retriable failure. 
- I deleted the service to try again. This has been pushed in as a delta again (while the api server has already deleted the service from etcd and hence querying the apiserver after this would always return service not found)
- Now my FIFO queue looks like this  (Add Service) <- (Delete Service)
- Since my Add Service (CreateTCPLoadbalancer) always relies on the service object (in my cloud provider) in the apiserver, this call would fail every single time hence forth. And is pushed back to the queue everytime (retriable is True) and the delete event does not get processed at all. 

I was able to reproduce this by inducing an error in the createTCPLB call everytime. I know there can be several ways to handle this but wanted to check with you guys first if my analysis on this problem makes any sense.

**Comment by brendandburns** [comment]
@lavalamp 

1) We should distinguish between transient failures and permanent failures in CreateTCPBalancer, there are lots of reasons that a failure might be persistent (over quota, invalid IP, etc) and we shouldn't retry those indefinitely.

2) We should perhaps change the semantics of the FIFO queue so that an indefinitely failing message can't block up an entire queue, that seems bad.

**Comment by uruddarraju** [comment]
https://groups.google.com/forum/#!topic/google-containers/-0tL-_lOV-Q

**Comment by lavalamp** [comment]
I'm still reading this, but: you should just read (i.e. list/watch) the endpoints object; there's an entire controller that does this query for you, efficiently.

**Comment by lavalamp** [comment]
Wait, I didn't understand the situation correctly. Let me take a look at the service controller code and get back.

**Comment by lavalamp** [comment]
@brendandburns it doesn't block the entire queue, it just blocks progress on a particular object.

**Comment by uruddarraju** [comment]
We do use the endpoints controller, but the CreateTCP call is only called within servicecontroller. But the updates to the LB when replication factor changes, or when pods are replaced, goes through the endpoint controller.

**Comment by lavalamp** [comment]
OK there are multiple problems with the service controller.

a. The IP addresses it gives are that of all our nodes. It should be configurable to get the list of endpoints, which is what sane load balancer implementations will want.
 a. It should get the nodes by listing & watching, instead of listing every time.
b. A crash or restart of the controller can leak load balancers; nowhere does it read the existing load balancers from the cloud provider.
c. A comment says "No delta compressor is needed for the DeltaFIFO queue because we only ever care about the most recent state" but that is inconsistent; you'd have to supply a compressor to throw away the entries you don't care about. (This is probably the cause of OP's bug)
d. A workqueue should be used so that we can process multiple services in parallel.

I will try to patch c above, since it seems like a bug. I think a and b are also pretty important but may need to wait until after 1.0.

**Comment by uruddarraju** [comment]
Thanks for your analysis @lavalamp. The reason for me to initiate this discussions is to find the right patterns. Can I try the (c) part and submit a PR and get some feedback if this is not something that has to to be dealt with immediate urgency (Next 1/2 days) ?

**Comment by lavalamp** [comment]
@uruddarraju I actually have most of a fix already. :) I'll toss it up in a PR shortly & you can see if it resolves your issue. I can't promise it'll be accepted for 1.0.

**Comment by davidopp** [comment]
At master team meeting we decided to move this to post-1.0 because it only blocks one object, not the whole queue.

**Comment by a-robinson** [comment]
This is still causing some load balancer leakages in e2e tests, as @cjcullen has confirmed this afternoon. One of us should tackle it soon. @lavalamp are you still on it, or should one of us take over?

**Comment by lavalamp** [comment]
My plate is pretty full, it's all yours :)

**Comment by a-robinson** [comment]
This was fixed before 1.2 by #22069 and #21982
